### PR TITLE
[WIP] Reference errors (attempt 2)

### PR DESCRIPTION
**Pull request by rochoa**
cc @donflopez 

This is a WIP PR, we still need to work on a proper solution that does not mix types: like the float vs expression case.

### CODE REVIEW

**Comment by donflopez** [comment]
[Original issue](https://github.com/CartoDB/carto/issues/22)

**Comment by dgaubert** [comment]
FYI: I've just tested it in Node v6.9.5 and it worked. ;)

File: lib/carto/renderer_js.js [hunk 1]
```diff
@@ -244,6 +238,13 @@ CartoCSS.prototype = {
       var done = {};
       for(var i = 0; i < defs.length; ++i) {
         var def = defs[i];
+
+        def.toXML(parse_env, {});
```
Comment (dgaubert) [inline lib/carto/renderer_js.js]: Why do we need to check errors for def and not for `rule`?

Comment (donflopez) [inline lib/carto/renderer_js.js]: At the definition level they are completely built with the toJS() method, if you use toXML at rule level you're gonna have a type error, taking all the rules as expressions.

**Comment by rochoa** [comment]
Offline discussion: make this check optional, so old clients keep the same behaviour, and new clients can, optionally, request for the validation.

File: lib/carto/renderer_js.js [hunk 2]
```diff
@@ -235,6 +238,13 @@ CartoCSS.prototype = {
...
+
+        if (this.options.strict && parse_env.errors.message) {
```
Comment (donflopez) [inline lib/carto/renderer_js.js]: Could this be the option @rochoa?

Comment (rochoa) [inline lib/carto/renderer_js.js]: True! But, wouldn't be better to perform the `def.toXML` when the option is provided?

Comment (donflopez) [inline lib/carto/renderer_js.js]: Yes! I'll change it :)

**Comment by donflopez** [comment]
Done!

File: test/editor/editor.test.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+var ref = require('./reference.json');
+var carto = require('../../lib/carto/index.js');
+
+
+var ccss = new carto.RendererJS({
+    reference: ref,
+    strict: true
+}).render("#layer { polygon-fill: #abcdef; }");
```
Comment (alonsogarciapablo) [inline test/editor/editor.test.js]: What are we testing here? Is this test being run?

Comment (donflopez) [inline test/editor/editor.test.js]: It looks like a residual file

Comment (alonsogarciapablo) [inline test/editor/editor.test.js]: :fire:?

File: lib/carto/renderer_js.js [hunk 4]
```diff
@@ -202,12 +202,15 @@ CartoCSS.prototype = {
         var layer = layers[key] = (layers[key] || {
           symbolizers: []
         });
+
         for(var u = 0; u<def.rules.length; u++){
-            if(def.rules[u].name === "marker-file" || def.rules[u].name === "point-file"){
-                var value = def.rules[u].value.value[0].value[0].value.value;
+          var rule = def.rules[u];
```
Comment (alonsogarciapablo) [inline lib/carto/renderer_js.js]: This line is not indented correctly.

File: test/editor/editor.test.js (hunk 3 above)
Comment (donflopez) [inline test/editor/editor.test.js]: 💣

File: test/rendering_js-strict.test.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    return function () {
+      var shader = RendererJS.render(style);
+
+      assert.ok(shader.layers);
+      assert.equal(shader.layers.length, 2);
+    };
+  }
+
+  it('should pass if a feature is not supported but strict mode is not specified', rendererStrictModeOffTest(
+    new carto.RendererJS({reference: reference, mapnik_version: '1.0.0' })
```
Comment (alonsogarciapablo) [inline test/rendering_js-strict.test.js]: could we use [assert.throws](https://www.npmjs.com/package/assert#assertthrowsblock-error-message) here and in the following test)?

File: lib/carto/renderer_js.js (hunk 4 above)
Comment (donflopez) [inline lib/carto/renderer_js.js]: I think that are the next lines what are bad indented

File: test/rendering_js-strict.test.js (hunk 5 above)
Comment (donflopez) [inline test/rendering_js-strict.test.js]: for what? here there are not any throw?

File: lib/carto/renderer_js.js (hunk 4 above)
Comment (alonsogarciapablo) [inline lib/carto/renderer_js.js]: True.

File: test/rendering_js-strict.test.js (hunk 5 above)
Comment (alonsogarciapablo) [inline test/rendering_js-strict.test.js]: (IMO) We should assert that it doesn't throw an error. `assert.doesNotThrow` would be  👌

Comment (donflopez) [inline test/rendering_js-strict.test.js]: assert.doesNotThrow seems to be created to check that when something throws an exception, it is not of a specific type.

If you use it without throwing, the test finish with those task as pending.

Comment (donflopez) [inline test/rendering_js-strict.test.js]: IMO, it not makes sense check that something does not throw, if the throw himself is gonna mark the task as not passed.

Comment (alonsogarciapablo) [inline test/rendering_js-strict.test.js]: :+1:

**Comment by alonsogarciapablo** [comment]
:es:

**Pull request merged**